daffodil-lib/src/main/scala/org/apache/daffodil/util/Validators.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.util
+
+import java.util.ServiceLoader
+
+import org.apache.daffodil.api.CompiledValidator
+import org.apache.daffodil.api.Validator
+import org.apache.daffodil.api.Validator.CompilerOps.CheckArgs
+import org.apache.daffodil.api.ValidatorNotFoundException
+
+import scala.collection.JavaConverters._
+
+object Validators {
+  val default: Validator = new DefaultValidatorSPIProvider
+  def compiler(): ValidatorCompiler = new Compiler
+
+  trait ValidatorCompiler {
+    def find(name: String): Option[CompiledValidator]
+    def isRegistered(name: String): Boolean
+    def compile(name: String, args: Validator.ArgumentList): CompiledValidator
+  }
+
+  private class Compiler extends ValidatorCompiler {
+    private var compiled = Map.empty[String, CompiledValidator]
+    private lazy val impls: Map[String, Validator] =
+      ServiceLoader
+        .load(classOf[Validator])
+        .iterator()
+        .asScala
+        .map(v => v.name() -> v)
+        .toMap
+
+    private def impl(name: String): Option[Validator] = impls.get(name)
+    def find(name: String): Option[CompiledValidator] = compiled.get(name)
+    def isRegistered(name: String): Boolean = impls.contains(name)
+
+    def compile(name: String, args: Validator.ArgumentList): CompiledValidator = {
+      find(name) match {
+        case Some(cv) => cv
+        case None =>
+          val vv = impl(name).getOrElse(throw ValidatorNotFoundException(name))
+          val aa = vv match {
+            case ac: CheckArgs => ac.checkArgs(args)
+            case _             => Validator.Arguments.Empty
+          }
+          val cv = new CompiledValidator(vv, aa)
```

Comment:
Here is where you want to call the specific compilation method for this specific provider, which will return a validator-specific object that implements the CompiledValidator trait.